torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -141,6 +141,67 @@ bool isSupported(Node* node) {
   return false;
 }
 
+// Code to detect specializations for testing CustomPasses.
+static std::unique_ptr<RegisterPass> tTSpecializationDetectionPass = nullptr;
+static bool hasSpecializations = false;
+
+bool hasTensorTypeSpecialization(Value* v) {
+  if (!v->type()->cast<TensorType>()) {
+    return false;
+  }
+  // Constants & TensorExprGroup will always produce specialized tensor type,
+  // TypeCheck are inserted by this pass and only used by fusion groups that
+  // insert proper guards
+  if (v->node()->kind() == prim::Constant ||
+      v->node()->kind() == prim::TypeCheck ||
+      v->node()->kind() == prim::TensorExprGroup) {
+    return false;
+  }
+  if (v->type() == TensorType::get()) {
+    return false;
+  }
+  return true;
+}
+
+bool hasTensorTypeSpecializations(torch::jit::Block* block) {
+  for (Value* v : block->inputs()) {
+    if (hasTensorTypeSpecialization(v))
+      return true;
+  }
+  for (Node* n : block->nodes()) {
+    for (torch::jit::Block* b : n->blocks()) {
+      if (hasTensorTypeSpecializations(b))
+        return true;
+    }
+    for (Value* v : n->outputs()) {
+      if (hasTensorTypeSpecialization(v))
+        return true;
+    }
+  }
+  return false;
+}
+
+void detectTTSpecializationPass(std::shared_ptr<Graph>& graph) {
```

Comment:
All of this code should be registered as a custom pass in a C++ test file, instead of having the test leak into our implementation